### PR TITLE
[WIP] add Alpha tags for alpha band in Uint16 One band dataset

### PR DESCRIPTION
**Pull request by vincentsarago**
## What does this PR do?
Make sure to add an AlphaMask flag for alpha band in a Uint16 1 band dataset. 

## What are related issues/pull requests?
https://lists.osgeo.org/pipermail/gdal-dev/2018-July/048832.html

## Tasklist
 - [ ] Add test case
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

**Comment by rouault** [comment]
Looks good. Could you add a test for that in autotest/gcore/mask.py ?  Taking inspiration from mask_24 should help

**Comment by vincentsarago** [comment]
@rouault I added a test case (thanks for pointing out where to look 👼 ).

**Comment by rouault** [comment]
```
  TEST: mask_26 ... fail (blowup)
Traceback (most recent call last):
  File "pymod/gdaltest_python2.py", line 43, in run_func
    result = func()
  File "./mask.py", line 1158, in mask_26
    if struct.unpack('B', mask.ReadRaster(0, 0, 1, 1))[0] != 255:
UnboundLocalError: local variable 'mask' referenced before assignment
Temporary /vsimem/ files remaining: ['mask_26.tif']
```

File: autotest/gcore/mask.py [hunk 1]
```diff
@@ -1137,6 +1137,50 @@ def mask_25():
...
+    mask = ds.GetRasterBand(1).GetMaskBand()
+
+    # IRasterIO() optimized case
+    import struct
+    if struct.unpack('B', mask.ReadRaster(0, 0, 1, 1))[0] != 255:
+        gdaltest.post_reason('fail')
+        print(struct.unpack('B', mask.ReadRaster(0, 0, 1, 1))[0])
+        return 'fail'
+
+    # IReadBlock() code path
```
Comment (rouault) [inline autotest/gcore/mask.py]: you could remove that block and the following one, as they test the same code paths as the RGBA case

File: autotest/gcore/mask.py [hunk 2]
```diff
@@ -1151,25 +1151,12 @@ def mask_26():
     if ds.GetRasterBand(1).GetMaskFlags() != gdal.GMF_ALPHA + gdal.GMF_PER_DATASET:
         gdaltest.post_reason('Did not get expected mask.')
         return 'fail'
-    mask = ds.GetRasterBand(1).GetMaskBand()
-
-    # IRasterIO() optimized case
```
Comment (rouault) [inline autotest/gcore/mask.py]: ah, this case  "#IRasterIO() optimized case" was the interesting one to let. The "special case where dynamics is only 0-255" is not so interesting

**Comment by vincentsarago** [comment]
@rouault sorry to restarting this discussion but I'm hitting the same problem again (1band, 16bytes dataset and internal mask) but this time my datasets are either in `Int16` or in `float64` so not caught by the `if` tests for `GDT_UInt16` and `GDT_Byte` 

I wonder if you have any though on this 

cc @sgillies

**Comment by rouault** [comment]
The semantics of an alpha band with Int16 or Float64 data type is unclear to me. The semantics of the mask band returned with GetMaskBand() is according to https://trac.osgeo.org/gdal/wiki/rfc15_nodatabitmask : "The masks are represented as GDT_Byte bands with a value of zero indicating nodata and non-zero values indicating valid data."

**Comment by vincentsarago** [comment]
I'm sorry to have mentioned internal mask. My problem is when using rasterio WarpedVRT https://github.com/mapbox/rasterio/blob/master/rasterio/_warp.pyx#L602 I think it created a virtual dataset with an alpha band to reflect the internal mask value. 

When checking the GDAL code we updated, I ended up thinking that an alpha won't be fetched in case the dataset is not either in `GDT_UInt16` and `GDT_Byte`

https://github.com/OSGeo/gdal/blob/772464bbc231e2a19236e217359940935e52979a/gdal/gcore/gdalrasterband.cpp#L6066-L6091


Which is what I'm getting when working a Float32 or Int16 dataset. This might be on rasterio side tho

**Comment by vincentsarago** [comment]
@rouault I started some test using https://github.com/vincentsarago/gdal/commit/496f78e47e57a171073cd9f11f991449df49cc84

Basically this commit allows alpha band masking for any datatype, but :
- this solution seems way to simple, so I'd love to have your though on it before submitting a PR 
- not sure how Complex datatype will be handle by `GDALRescaledAlphaBand`

Note, while this seems to work (alpha band read correctly) there seems to be a huge performance downgrade.

Benchmark done https://github.com/vincentsarago/benchmark-vrt-datatype#benchmark-vrt

The performance issue might be on Rasterio VRT+alpha side but I'll need to run some tests 

cc @sgillies

**Pull request merged**